src/emoji/data.js

```diff
@@ -1,35 +1,141 @@
+const refreshServerEmojiData = (data: ServerEmojiData): PerAccountAction => ({
+  type: REFRESH_SERVER_EMOJI_DATA,
+  data,
+});
 
-const unicodeEmojiObjects: $ReadOnlyArray<EmojiForShared> = objectEntries(unicodeCodeByName).map(
-  ([name, code]) => ({
-    emoji_type: 'unicode',
-    emoji_name: name,
-    emoji_code: code,
-  }),
-);
+export const maybeRefreshServerEmojiData =
+  (server_emoji_data_url: string | void): ThunkAction<Promise<void>> =>
+  async (dispatch, getState) => {
+    if (server_emoji_data_url === undefined) {
+      // The server is too old to support this feature.
+      // TODO(server-6.0): Simplify away; should always be present.
+      return;
+    }
+    const parsedUrl = tryParseUrl(server_emoji_data_url);
+    if (!parsedUrl) {
+      logging.error('Invalid URL at server_emoji_data_url in /register response');
+      return;
+    }
+
+    let data = undefined;
+    try {
+      data = await api.fetchServerEmojiData(parsedUrl);
+    } catch (errorIllTyped) {
+      const e: mixed = errorIllTyped; // https://github.com/facebook/flow/issues/2470
+
+      if (!(e instanceof Error)) {
+        logging.error('Unexpected non-error thrown from api.fetchServerEmojiData.');
+        return;
+      }
+
+      if (e instanceof NetworkError) {
+        // Probably a user-connectivity error; no need to log. Hope it goes
+        // through next time.
+        // TODO(?): Could retry?
+      } else {
+        // Non-network errors might signal a server or client bug; log them.
+        logging.error(e);
+      }
+      return;
+    }
+
+    dispatch(refreshServerEmojiData(data));
+  };
+
+const getUnicodeEmojiObjects = (
+  serverEmojiData: ServerEmojiData | null,
+): $ReadOnlyArray<EmojiForShared> => {
+  if (!serverEmojiData) {
+    return objectEntries(unicodeCodeByName).map(([name, code]) => ({
+      emoji_type: 'unicode',
+      emoji_name: name,
+      emoji_code: code,
+    }));
+  }
+
+  const result = [];
+  for (const [code, names] of serverEmojiData.code_to_names.entries()) {
+    result.push(
+      ...names.map(name => ({
+        emoji_type: 'unicode',
+        emoji_name: name,
+        emoji_code: code,
+      })),
+    );
+  }
+  return result;
+};
+
+const getUnicodeEmojiObjectsMemoized = defaultMemoize(getUnicodeEmojiObjects);
 
-export const parseUnicodeEmojiCode = (code: string): string /* force line */ =>
+/**
+ * Convert a Unicode emoji's `emoji_code` into the actual Unicode codepoints.
+ *
+ * Follows the comment on emoji_code in
+ *   https://github.com/zulip/zulip/blob/main/zerver/models.py :
+ *
+ * > * For Unicode emoji, [emoji_code is] a dash-separated hex encoding of
+ * >   the sequence of Unicode codepoints that define this emoji in the
+ * >   Unicode specification.  For examples, see "non_qualified" or
+ * >   "unified" in the following data, with "non_qualified" taking
+ * >   precedence when both present:
+ * >   https://raw.githubusercontent.com/iamcal/emoji-data/master/emoji_pretty.json
```

Comment:
nit: this part is really about the implementation, not interface, so should be in a plain implementation comment rather than jsdoc.

(In particular, the caller should not have to care about the details of the emoji_code format -- that's this function's job.)